### PR TITLE
MS RFC 132: Update MapScript API Docs

### PR DESCRIPTION
**Pull request by geographika**
This pull request implements [RFC 132](https://mapserver.org/development/rfc/ms-rfc-132.html) - the (semi)-automatic generation of MapScript API documentation using the Python SWIG bindings. 

To see the generated Sphinx output see [MapScript API Index](https://geographika.github.io/MapServer-documentation/mapscript/mapscript-api/index.html) and [MapScript Index Page](https://geographika.github.io/MapServer-documentation/mapscript/index.html). 

Features
---------

* Automatic stub generation based on the latest MapScript release. Keywords and functions. 
* Any undocumented properties or functions has a TODO label added. The docs will always be up to date with the latest Python MapScript release
* Linking to Mapfile keywords to avoid duplication of docs (see [Sphinx pull request](https://github.com/sphinx-doc/sphinx/pull/7087))
* Linking to classes and functions available throughout docs, e.g. in HowTo and in the API docs themselves.
* Support for Python snippets from the Python test cases, ensuring examples continue to work with the latest codebase
* Set of Python scripts to update the list of available classes, constants, and enumerations, as well as class diagrams. These could also be set these to run automatically as part of the CI / docs build process
* Updates to the various MapScript HOWTO guides

Update Process
----------------

- On each MapServer version release Appveyor [automatically publishes](https://github.com/MapServer/MapServer/blob/3ddcf5111481f41821eb020f155ab0631331cdb0/appveyor.yml#L128) a source distribution of MapScript to PyPI
- The MapServer CI Travis script then downloads the distribution as part of the Sphinx documentation build. MapServer *is not required* to be installed - mocks are used so only Python .py files are required
- Various Sphinx extensions take care of creating pages for each of the MapScript classes

TODO
------

**Note** the Travis script is currently downloading mapscript from [Test PyPI](https://test.pypi.org/project/mapscript/7.7.4/). A custom SWIG build was used to generate this as it allows for property annotations to be added automatically to the output. There is an open [pull request](https://github.com/swig/swig/pull/1951) in SWIG to add this (requiring some unit tests). Once this is added to SWIG mapscript can be downloaded from standard PyPI and added to requirements.txt

In `conf.py` `nitpicky = True` has been temporarily set to `nitpicky = False`. Many of the `py:obj` references were throwing errnoneous errors. These can be turned off using the following setting, but this is only available in the as-yet unrelease 4.1 version of Sphinx:

```
nitpick_ignore_regex = [('py:obj', 'mapscript.*')]](https://github.com/MapServer/MapServer/pull/ New in Sphinx version 4.1
```

 Associated Pull Requests
--------------------------

Along with this documentation pull request, there are a number of updates to the MapServer sourcecode to add documentation strings along with the code:

* [Update doxygen docstrings](https://github.com/MapServer/MapServer/pull/6337)
* [Mapscript legacy functions cleanup](https://github.com/MapServer/MapServer/pull/6332)
* [Document various MapServer structs](https://github.com/MapServer/MapServer/pull/6248)
* [Document webObj, styleObj and referenceMapObj](https://github.com/MapServer/MapServer/pull/6245)
* [Document outputFormatObj, projectionObj and map primitives](https://github.com/MapServer/MapServer/pull/6244)
* [Document labelCacheObj and associated objects (RFC 132)](https://github.com/MapServer/MapServer/pull/6224)
* [Add labelObj and layerObj comments to mapserver.h](https://github.com/MapServer/MapServer/pull/6237)
* [Document legendObj and lineObj](https://github.com/MapServer/MapServer/pull/6238)
* [Update mapObj comments and docstring fixes](https://github.com/MapServer/MapServer/pull/6242)
* [Document imageObj and hashTableObj](https://github.com/MapServer/MapServer/pull/6168)
* [Document the DBFInfo and shapefileObj with Doxygen comments](https://github.com/MapServer/MapServer/pull/6152)
* [Add property docstrings to classObj](https://github.com/MapServer/MapServer/pull/6140)
* [Add docstrings to all MapScript classes and methods](https://github.com/MapServer/MapServer/pull/6118)
* [SWIG Documentation for colorObj](https://github.com/MapServer/MapServer/pull/6075)
* [SWIG Documentation for clusterObj](https://github.com/MapServer/MapServer/pull/6074)
* [Add classObj property strings and update MapScript docstrings](https://github.com/MapServer/MapServer/pull/6071)
* [Add docstrings to SWIG objects](https://github.com/MapServer/MapServer/pull/5930)
* [Generated Mapscript API Documentation](https://github.com/MapServer/MapServer/pull/5870)
* [Document labelObj properties](https://github.com/MapServer/MapServer/pull/6217)

### CODE REVIEW

File: conf.py [hunk 1]
```diff
@@ -80,10 +118,10 @@
 pygments_style = 'sphinx'
 
 # Exclude git directories
-exclude_patterns = ['.git', 'howto', 'redirection', 'users-manual']
+exclude_patterns = ['.git', 'howto', 'redirection', 'users-manual', 'mapscript/mapscript-api/constants', 'mapscript/mapscript-api/mapscript']
 
 # check for broken reference targets
-nitpicky = True
+nitpicky = False
```
Comment (rouault) [inline conf.py:124]: why this change ?

Comment (geographika) [reply conf.py:124]: See TODO above. There are 100s of erroneous warnings such as `WARNING: py:obj reference target not found: mapscript.symbolObj.miny` in the stub files (even though the links work fine). Sphinx 4.1 has a way of filtering these out but is not yet released.
